### ISSUE TITLE
Does SecureChannel.GenerateToken need to be blocking the IO threads?

### ISSUE THREAD

**Issue by TimLovellSmith** [dotnet/runtime#39542]
In our .net framework app, I notice that `SecureChannel.GenerateToken` blocks the IO thread for some time as part of SSL negotiation - I think this is probably part of the `async` callback series we are doing for `SslStream.AuthenticateAsClientAsync` or `SslStream.AuthenticateAsServerAsync`, right?

Is there some way we could get this to not block the thread? Could the framework use `SspiInitializeSecurityContextAsync` or `SspiAcceptSecurityContextAsync` functions, and then not need to block? 

I guess there are some other sync method signatures in between which would need async versions as well, including SecureChannel.GenerateToken itself, but hopefully it would be a simple sync to async refactoring...

Or are there some better APIs than these to use?

```
ntoskrnl.exe!KiSwapContext
ntoskrnl.exe!KiSwapThread
ntoskrnl.exe!KiCommitThreadWait
ntoskrnl.exe!KeWaitForSingleObject
ntoskrnl.exe!AlpcpSignalAndWait
ntoskrnl.exe!AlpcpReceiveSynchronousReply
ntoskrnl.exe!AlpcpProcessSynchronousRequest
ntoskrnl.exe!NtAlpcSendWaitReceivePort
ntoskrnl.exe!KiSystemServiceCopyEnd
ntdll.dll!ZwAlpcSendWaitReceivePort
rpcrt4.dll!LRPC_CCALL::SendReceive
rpcrt4.dll!NdrpClientCall3
rpcrt4.dll!NdrClientCall3
sspicli.dll!SspipProcessSecurityContext
sspicli.dll!LsaInitializeSecurityContextCommon
sspicli.dll!LsaInitializeSecurityContextW
sspicli.dll!InitializeSecurityContextCommon
sspicli.dll!InitializeSecurityContextW
System.ni.dll!DomainBoundILStubClass.IL_STUB_PInvoke(System.Net.SSPIHandle ByRef, Void*, Byte*, System.Net.ContextFlags, Int32, System.Net.Endianness, System.Net.SecurityBufferDescriptor, Int32, System.Net.SSPIHandle ByRef, System.Net.SecurityBufferDescriptor, System.Net.ContextFlags ByRef, Int64 ByRef)$##6000000
System.ni.dll!System.Net.SafeDeleteContext.MustRunInitializeSecurityContext_SECURITY(System.Net.SafeFreeCredentials ByRef, Void*, Byte*, System.Net.ContextFlags, System.Net.Endianness, System.Net.SecurityBufferDescriptor, System.Net.SafeDeleteContext, System.Net.SecurityBufferDescriptor, System.Net.ContextFlags ByRef, System.Net.SafeFreeContextBuffer)$##6001321
System.ni.dll!System.Net.SafeDeleteContext.MustRunInitializeSecurityContext_SECURITY(System.Net.SafeFreeCredentials ByRef, Void*, Byte*, System.Net.ContextFlags, System.Net.Endianness, System.Net.SecurityBufferDescriptor, System.Net.SafeDeleteContext, System.Net.SecurityBufferDescriptor, System.Net.ContextFlags ByRef, System.Net.SafeFreeContextBuffer)$##6001321
System.ni.dll!System.Net.SafeDeleteContext.InitializeSecurityContext(System.Net.SecurDll, System.Net.SafeFreeCredentials ByRef, System.Net.SafeDeleteContext ByRef, System.String, System.Net.ContextFlags, System.Net.Endianness, System.Net.SecurityBuffer, System.Net.SecurityBuffer[], System.Net.SecurityBuffer, System.Net.ContextFlags ByRef)$##6001320
System.ni.dll!System.Net.SSPISecureChannelType.InitializeSecurityContext(System.Net.SafeFreeCredentials, System.Net.SafeDeleteContext ByRef, System.String, System.Net.ContextFlags, System.Net.Endianness, System.Net.SecurityBuffer[], System.Net.SecurityBuffer, System.Net.ContextFlags ByRef)$##60011E5
System.ni.dll!System.Net.SSPIWrapper.InitializeSecurityContext(System.Net.SSPIInterface, System.Net.SafeFreeCredentials, System.Net.SafeDeleteContext ByRef, System.String, System.Net.ContextFlags, System.Net.Endianness, System.Net.SecurityBuffer[], System.Net.SecurityBuffer, System.Net.ContextFlags ByRef)$##6001390
System.ni.dll!System.Net.Security.SecureChannel.GenerateToken(Byte[], Int32, Int32, Byte[] ByRef)$##6001E53
System.ni.dll!System.Net.Security.SecureChannel.NextMessage(Byte[], Int32, Int32)$##6001E52
System.ni.dll!System.Net.Security.SslState.StartSendBlob(Byte[], Int32, System.Net.AsyncProtocolRequest)$##6001F63
System.ni.dll!System.Net.Security.SslState.ProcessReceivedBlob(Byte[], Int32, System.Net.AsyncProtocolRequest)$##6001F67
System.ni.dll!System.Net.Security.SslState.ReadFrameCallback(System.Net.AsyncProtocolRequest)$##6001F6C
System.ni.dll!System.Net.AsyncProtocolRequest.CompleteRequest(Int32)$##60013FC
System.ni.dll!System.Net.FixedSizeReader.CheckCompletionBeforeNextRead(Int32)$##60013EF
System.ni.dll!System.Net.FixedSizeReader.ReadCallback(System.IAsyncResult)$##60013F0
System.ni.dll!System.Net.LazyAsyncResult.Complete(IntPtr)$##6001154
mscorlib.ni.dll!System.Threading.ExecutionContext.RunInternal(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object, Boolean)$##6003AD4
mscorlib.ni.dll!System.Threading.ExecutionContext.Run(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object, Boolean)$##6003AD3
mscorlib.ni.dll!System.Threading.ExecutionContext.Run(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object)$##6003AD2
System.ni.dll!System.Net.ContextAwareResult.Complete(IntPtr)$##60010B2
System.ni.dll!System.Net.LazyAsyncResult.ProtectedInvokeCallback(System.Object, IntPtr)$##6001151
System.ni.dll!System.Net.Sockets.BaseOverlappedAsyncResult.CompletionPortCallback(UInt32, UInt32, System.Threading.NativeOverlapped*)$##60021FB
mscorlib.ni.dll!System.Threading._IOCompletionCallback.PerformIOCompletionCallback(UInt32, UInt32, System.Threading.NativeOverlapped*)$##6003B52
```

**Comment by wfurt** [comment]
Bar for .net framework  is really really high. It is unlikely we would do any major changes at this point. 

The bottom is all synchronous and moving it to different thread would come with performance and complexity penalty IMHO. The NextMessgae/GenerateToken as still somewhat trivial but the [InitializeSecurityContext](https://docs.microsoft.com/en-us/windows/win32/api/sspi/nf-sspi-initializesecuritycontexta) already represents OS function. 

The current .NET Core strategy is to use Async for around IO as that can take a while and needs to be queued but then marshal everything via Span and process as much as quickly as we can.

**Comment by TimLovellSmith** [comment]
@wfurt 
Scenario - I looked at captured profile data from the cloud. In production, in aggregate, all calls to `SslState.ReadFrameCallback` from IO threads were spending about 81% of their total execution time blocked on this one API.

That factoid wouldn't worry me as much if it were not happening on the IO threads. The worry is that by blocking these it will affect the latency of IO handling for other established connections.

I guess what I'd like to achieve is effectively that new connection handling isn't blocking established connection IO processing, i.e. treat it as lower priority.

Any ideas on a good way to achieve that?
Could using a custom SychronizationContext for our SslStream calls, so that it wouldn't actually run LazyAsyncResult on the IO threads, work in theory?

**Comment by karelz** [comment]
Triage: Not important for 5.0, we may look into it later, but we may also decide it is not worth it on .NET Core.
It would require some perf numbers.

Workaround: Increase number of IO threads.

**Comment by wfurt** [comment]
Looking at this again, I don't think this is really actionable. This would need major re-work and I don't think we can justify the risk.

**Comment by TimLovellSmith** [comment]
@wfurt I'm not planning to argue hard for this, I'm just thinking I should point our that the worst impact I can think of here is the scenarios here verge on DOS-vulnerability, where excessive SSL negotiation thwarts your ability to handle other established network connection. There may be other reasonable ways to mitigate that... maybe...

**Comment by wfurt** [comment]
Framework it basically frozen in time  but things get easier in .NET (Core) You _could_ make custom Streams and layer custom logic in if you want to. If you care about perf and scalability it is way to go anyway @TimLovellSmith
